components/script/textinput.rs

```diff
@@ -1128,4 +1137,79 @@ impl<T: ClipboardProvider> TextInput<T> {
             .fold(UTF8Bytes::zero(), |acc, x| acc + x.len_utf8());
         self.edit_point.index = byte_offset;
     }
+
+    fn paste_contents(&mut self, drag_data_store: &DragDataStore) {
+        for item in drag_data_store.iter_item_list() {
+            if let Kind::Text(string) = item {
+                self.insert_string(string.data());
+            }
+        }
+    }
+}
+
+pub(crate) fn handle_text_clipboard_action(
+    owning_node: &impl NodeTraits,
+    textinput: &DomRefCell<TextInput<ScriptToConstellationChan>>,
+    event: &ClipboardEvent,
+    can_gc: CanGc,
+) -> bool {
+    let e = event.upcast::<Event>();
+
```

Comment:
The spec doesn't have an explicit step for this, but we need to check the IsTrusted value of the event and prevent any changes to the system clipboard if it's false.